cloud/pkg/admissioncontroller/controller/admission.go

```diff
@@ -0,0 +1,273 @@
+package controller
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/kubeedge/kubeedge/cloud/pkg/admissioncontroller/constants"
+	"github.com/kubeedge/kubeedge/cloud/pkg/admissioncontroller/utils"
+	devicesv1alpha1 "github.com/kubeedge/kubeedge/cloud/pkg/apis/devices/v1alpha1"
+	devivicecontrollerutils "github.com/kubeedge/kubeedge/cloud/pkg/devicecontroller/utils"
+
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	"k8s.io/utils/pointer"
+)
+
+var scheme = runtime.NewScheme()
+var codecs = serializer.NewCodecFactory(scheme)
+
+func init() {
+	addToScheme(scheme)
+}
+
+func addToScheme(scheme *runtime.Scheme) {
+	utilruntime.Must(corev1.AddToScheme(scheme))
+	utilruntime.Must(admissionv1beta1.AddToScheme(scheme))
+	utilruntime.Must(admissionregistrationv1beta1.AddToScheme(scheme))
+	utilruntime.Must(devivicecontrollerutils.AddDeviceCrds(scheme))
+}
+
+// AdmissionController implements the admission webhook for validation of configuration.
+type AdmissionController struct {
+	Client *kubernetes.Clientset
+}
+
+func strPtr(s string) *string { return &s }
+
+// Start starts the webhook service
+func (ac *AdmissionController) Start(context *utils.CertContext) {
+	//TODO: read somewhere to get what's kind of webhook is enabled, register those webhook only.
+	err := ac.registerWebhookForDeviceModel(constants.WebhookNameDeviceModel, context)
+	if err != nil {
+		klog.Fatalf("Failed to register the webhook with error: %v", err)
+	}
+	ac.deployService()
+	tlsConfig := utils.ConfigTLS(context)
+	http.HandleFunc("/devicemodels", serveDeviceModel)
+	server := &http.Server{
+		Addr:      fmt.Sprintf(":%v", constants.Port),
+		TLSConfig: tlsConfig,
+	}
+
+	server.ListenAndServeTLS("", "")
+}
+
+// Register registers the admission webhook.
+func (ac *AdmissionController) registerWebhookForDeviceModel(WebhookName string, context *utils.CertContext) error {
+	webhook := &admissionregistrationv1beta1.ValidatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: WebhookName,
+		},
+		Webhooks: []admissionregistrationv1beta1.ValidatingWebhook{
+			{
+				Name: WebhookName,
+				Rules: []admissionregistrationv1beta1.RuleWithOperations{{
+					Operations: []admissionregistrationv1beta1.OperationType{
+						admissionregistrationv1beta1.Create,
+						admissionregistrationv1beta1.Update,
+						admissionregistrationv1beta1.Delete,
+						admissionregistrationv1beta1.Connect,
```

Comment:
Does the webhook need catch delete and connect here? I think create and update is enough

Reply:
there is no action for delete or connect so far, I think there is no harm to add it there but might good for extension, we can remove them if you think it's not necessary.